Documentation/bpf.rst

```diff
@@ -0,0 +1,1855 @@
+key / value stores, helper functions to interact with the kernel, tail calls for
+calling into other eBPF programs, security hardening primitives, a pseudo file
+system for pinning objects (maps, programs), and infrastructure for allowing
+eBPF to be offloaded, for example, to a network card.
+
+LLVM provides an eBPF back end, such that tools like ``clang`` can be used to compile
+C into an eBPF object file, which can then be loaded into the kernel. eBPF is
+deeply tied into the Linux kernel and allows for full programmability without
+sacrificing native kernel performance.
+
+The following subsections provide further details on individual aspects of the
+eBPF architecture.
+
+Instruction Set
+---------------
+
+eBPF is a general purpose RISC instruction set and was originally designed with the
+goal to write programs in a subset of C that can be compiled into eBPF instructions
+through a compiler back end (e.g., LLVM), such that the kernel can later on map them
+through an in-kernel JIT compiler into native opcodes for optimal execution performance
+inside the kernel.
+
+The advantages for pushing these instructions into the kernel are, for example:
+
+* Making the kernel programmable without having to cross kernel / user space
+  boundaries. For example, eBPF programs related to networking as in the case of
+  Cilium, can implement flexible container policies, load balancing and other means
+  without having to move packets to user space and back into the kernel. State
+  between eBPF programs and kernel / user space can still be shared through maps
+  whenever needed.
+
+* Given the flexibility of a programmable data path, programs can be heavily optimized
+  for performance also by compiling out features that are not required for the use cases
+  the program solves. F.e., if a container does not require IPv4, then the eBPF program
+  can be built to only deal with IPv6 in order to save resources in the fast-path.
+
+* In case of networking (e.g., tc and XDP), eBPF programs can be updated atomically
+  without having to restart the kernel, system services or containers, and without
+  traffic interruptions.
+
+* eBPF provides a stable ABI towards user space, and does not require any third party
+  kernel modules, for example. eBPF is a core part of the Linux kernel that is shipped
+  everywhere, and guarantees that existing eBPF programs keep running with newer kernel
+  versions. This guarantee is the same guarantee that the kernel provides for system
+  calls with regard to user space applications.
+
+* eBPF programs work in concert with the kernel, they make use of existing kernel
+  infrastructure and tooling (e.g., iproute2) as well as the safety guarantees that
+  the kernel provides. Unlike kernel modules, eBPF programs are verified through an
+  in-kernel verifier in order to ensure that they cannot crash the kernel, always
+  terminate, etc. XDP programs, for example, reuse the existing in-kernel drivers
```

Comment:
Given it's kind of used throughout the doc (also in the more generic LLVM / iproute2 section), I tried to introduce XDP and tc terms at the very beginning now, but we could later on also still rearrange this a bit if we think a strict separation seems better.